Update submitter email on email update

kevinrobbins opened the pull request:
This PR solves the issue where users change their email address and lose access to their tickets. This was because tickets are associated to users by the email address tied to the User object. This PR addresses this by creating a signal in helpdesk that detects changes to the User object and updates the email address appropriately.

kevinrobbins committed:
Add signal to update submitter_email when User email changes

When a user changed their primary email address, they would lose access
to their tickets because they were associated with the old email
address. This signal updates their tickets to the new email address so
that the users will still be able to access their tickets.

kevinrobbins committed:
Add default app config

We need to be able to override the ready() method on the AppConfig in
order to register our signals module with Django.

kevinrobbins committed:
Handle new user creation in email_address_update signal

Because this is a pre_save signal, if a user is about to be created,
there is no user to query for and an exception is thrown. Here we change
the get() to a filter() and check to see if a user was returned. If no
user is found, then we can just return because they would not have any
tickets to update.

kevinrobbins committed:
Rename email_address_update signal

The name update_ticket_submitter_email more accurately describes what
the signal receiver does.

kevinrobbins committed:
Bump helpdesk version to 0.2.6+nimbis.22.1

We can't use 0.2.6+nimbis.23 because that version has already been
created in preparation for the django upgrade. We want to release the
site under django 1.11 which is not compatible with 0.2.6+nimbis.23.

kevinrobbins commented:
When merging, just take the version in setup.py in master. This version is just so we can release this pre-django upgrade.

kevinrobbins commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[tsscom-1147] Changed Email Address and now all old tickets are Forbidden (Opened)](https://trello.com/c/4sxKiAE1/10163-tsscom-1147-changed-email-address-and-now-all-old-tickets-are-forbidden-opened)

kevinrobbins commented:
I'm not worried about the checks. They are failing because of flake8 errors and I don't want to go in and fix dozens of errors in a package we mostly don't maintain.

Commit:
Fix bug with attaching files when submitting follow-ups on tickets.

The following ISE was being thrown when users submitted a follow-up
response on a ticket and attaching any file:

Exception Type: FileNotFoundError at /support/tickets/<TICKET_ID>/update/
Exception Value: [Errno 2] No such file or directory:
'helpdesk/attachments/<SITE_ID-TICKET_ID>/...'

In this commit we switched to using S3 storage for attachments
in send_templated_mail:
https://github.com/nimbis/django-helpdesk/commit/7807b12a9c575d90364a67a18655ab87ed622da2

However, we did not apply that same change in the 'if six.PY3'
consequent. This path was not hit when were were running sites using
python2. Once we switched to python3, the bug fixed in the commit
above re-appeared.

Additionally, the latest version of django-helpdesk does not
include this if-then statement.

References:
https://github.com/django-helpdesk/django-helpdesk/issues/721
https://github.com/django-helpdesk/django-helpdesk/commit/b43e60875a3339c57a68a1405f6874d57a5c19a2#diff-798c61915598ee76b4cb2ee8be71c807

kevinrobbins committed:
Merge pull request #25 from nimbis/fix-bug-with-attaching-file

Fix bug with attaching files when submitting follow-ups on tickets.

abannerjee commented on helpdesk/signals.py:
```diff
@@ -0,0 +1,17 @@
+    tickets = Ticket.objects.filter(submitter_email=pre_save_user.email)
+    for ticket in tickets:
+        ticket.submitter_email = instance.email
+        ticket.save()
```
This signal will run whenever the `User` object is saved (whether the primary email changes or not). And when that happens, we are updating all tickets associated with that user each time.

In order to limit potential race condition issues, I recommend using `update_fields=['submitter_email']` in the `save()` method. Additionally, it might be good to also check if the email has changed prior to updating the `ticket`  (if the email hasn't changed, we would not need to do anything)

kevinrobbins replied:
This is a great idea. I'll go ahead and make this change.

abannerjee commented:
@kevinrobbins Reviewed the code and left one comment which I think should be addressed/looked-at before merging.

I also recommend just rebasing this against master and updating `setup.py` to version `0.2.6+nimbis.24`. This is a mess, because we'll still need to use version `0.2.6+nimbis.22.1` on the sites, but at least version `0.2.6+nimbis.24` will include these fixes + django2.2 changes.

kevinrobbins committed:
Exit update_ticket_submitter_email early if email address didn't change

We don't need to update the ticket if the email address didn't change so
we can exit early. Here we also condense the early exit logic so that it
all happens in a single, unbroken chunk.

kevinrobbins committed:
Specify submitter_email as only update_field

This will limit potential race conditions by only updating the
submitter_email field. This way if some other property on the ticket is
updated (for example by an admin) while the email address is being
updated, we don't lose any other data.

kevinrobbins commented:
@abannerjee I made the changes you suggested. Once you re-review, rebase and create a version 24.
